File: Assets/Scripts/Utility/DynamicArray2D.cs

using System;
using System.Collections.Generic;

namespace Utility
{
	public class DynamicArray2D<T> where T : class
	{
		public Vec2i origin;
		public Vec2i dim;
		public T[] data;
		
		public DynamicArray2D()
		{
			this.origin = new Vec2i(0, 0);
			this.dim = new Vec2i(0, 0);
			this.data = new T[0];
		}

		public void AABBNonDef(out Vec2i bl, out Vec2i tr)
		{
			bl = new Vec2i(dim.x,dim.y);
			tr = new Vec2i(0,0);

			for (int i = 0; i < dim.y; i++) {
				for (int j = 0; j < dim.x; j++) {

					if (data [indexOf(j,i)] != null) {

						if (j < bl.x){
							bl.x = j;
						}
						if (i < bl.y){
							bl.y = i;
						}

						if (j >= tr.x){
							tr.x = j + 1;
						}
						if (i >= tr.y){
							tr.y = i + 1;
						}
					}

				}
			}

			bl += origin;
			tr += origin;

			if (bl.x > tr.x) {
				tr.x = bl.x;
			}
			if (bl.y > tr.y) {
				tr.y = bl.y;
			}
		}
		
		public IEnumerable<T> TryGetOutsideInside(Vec2i outside, Vec2i inside, int range)
		{
			int rangeT = (range * 2) + 1;

			for (int i = 0; i < rangeT; i++) {
				for (int j = 0; j < rangeT; j++) {

					Vec2i temp = inside + new Vec2i(i - range, j - range);

					if (!((temp - outside) <= range)) {
						T temp2 = null;
						if (TryGet (temp.x, temp.y, ref temp2)) {
							yield return TryGet(temp.x, temp.y);
						}
					}
				}
			}
		}
		
		public IEnumerable<T> WithinRange(Vec2i inside, int range)
		{
			int rangeT = (range * 2) + 1;
			
			for (int i = 0; i < rangeT; i++) {
				for (int j = 0; j < rangeT; j++) {
					
					Vec2i temp = inside + new Vec2i(i - range, j - range);
					
					T temp2 = null;
					if (TryGet (temp.x, temp.y, ref temp2)) {
						yield return TryGet(temp.x, temp.y);
					}
				}
			}
		}

		private void GrowTopRight(Vec2i amount)
		{
			Vec2i newDim = dim + amount;
			T[] newData = new T[newDim.x * newDim.y];

			for (int i = 0; i < amount.y; i++) {
				for (int j = 0; j < newDim.x; j++) {
					newData[indexOf(j,newDim.y - i, newDim.x)] = null;
				}
			}

			for (int i = 0; i < dim.y; i++) {
				for (int j = 0; j < dim.x; j++) {
					newData[indexOf(j,i, newDim.x)] = data[indexOf(j,i, dim.x)];
				}
				for (int j = dim.x; j < newDim.x; j++) {
					newData[indexOf(j,i, newDim.x)] = null;
				}
			}

			data = newData;
			dim = newDim;
		}
		
		private void GrowBottomLeft(Vec2i amount)
		{
			
			Vec2i newDim = dim + amount;
			T[] newData = new T[newDim.x * newDim.y];
			
			for (int i = amount.y; i < newDim.y; i++) {
				for (int j = 0; j < amount.x; j++) {
					newData[indexOf(j,i, newDim.x)] = null;
				}
				for (int j = amount.x; j < newDim.x; j++) {
					newData[indexOf(j,i, newDim.x)] = data[indexOf(j - amount.x,i - amount.y, dim.x)];
				}
			}
			
			for (int i = 0; i < amount.y; i++) {
				for (int j = 0; j < newDim.x; j++) {
					newData[indexOf(j,i, newDim.x)] = null;
				}
			}

			origin -= amount;
			data = newData;
			dim = newDim;
		}

		public bool Contains(int x, int y)
		{
			return (x >= origin.x && x < dim.x + origin.x) && (y >= origin.y && y < dim.y + origin.y);
		}
		
		public bool Contains(Vec2i index)
		{
			return (index.x >= origin.x && index.x < dim.x + origin.x) && (index.y >= origin.y && index.y < dim.y + origin.y);
		}
		
		public int indexOf (Vec2i index)
		{
			return (index.y * dim.x) + index.x;
		}
		
		public static int indexOf (Vec2i index, int columns)
		{
			return (index.y * columns) + index.x;
		}
		
		public int indexOf (int x, int y)
		{
			return (y * dim.x) + x;
		}
		
		public static int indexOf (int x, int y, int columns)
		{
			return (y * columns) + x;
		}
		
		public void Set (int x, int y, T val)
		{
			if (x < origin.x) {
				if (y < origin.y) {
					//expand bottom left
					GrowBottomLeft(new Vec2i(origin.x - x, origin.y - y));
				} else if (y >= dim.y + origin.y) {
					//expand top left
					GrowTopRight(new Vec2i(0, y - dim.y + origin.y + 1));
					GrowBottomLeft(new Vec2i(origin.x - x, 0));
				} else {
					//expand left
					GrowBottomLeft(new Vec2i(origin.x - x, 0));
				}
			} else if (x >= dim.x + origin.x) {
				if (y < origin.y) {
					//expand bottom right
					GrowTopRight(new Vec2i(x - dim.x + origin.x + 1, 0));
					GrowBottomLeft(new Vec2i(0, y - dim.y + origin.y + 1));
				} else if (y >= dim.y + origin.y) {
					//expand top right
					GrowTopRight(new Vec2i(x - dim.x + origin.x + 1, y - dim.y + origin.y + 1));
				} else {
					//expand right
					GrowTopRight(new Vec2i(x - dim.x + origin.x + 1, 0));
				}
			} else {
				if (y < origin.y) {
					//expand bottom
					GrowBottomLeft(new Vec2i(0, origin.y - y));
				} else if (y >= dim.y + origin.y) {
					//expand top
					GrowTopRight(new Vec2i(0, y - dim.y + origin.y + 1));
				}
				else {
					x -= origin.x;
					y -= origin.y;
					data [indexOf(x,y)] = val;
				}
			}
		}
		
		public bool TryGet (int x, int y, ref T val)
		{
			if (!Contains(x, y))
				return false;
			
			x -= origin.x;
			y -= origin.y;
			val = data [indexOf(x,y)];

			return true;
		}
		
		public T TryGet (int x, int y)
		{
			if (!Contains(x, y))
				return null;

			x -= origin.x;
			y -= origin.y;
			return data [indexOf(x,y)];
		}
		
		public T TryGet (Vec2i index)
		{
			if (!Contains(index.x, index.y))
				return null;
			
			index -= origin;
			return data [indexOf(index)];
		}
		
		public T Get (int x, int y)
		{
			x -= origin.x;
			y -= origin.y;
			return data [indexOf(x,y)];
		}
		
		public T Get2 (int x, int y)
		{
			return data [indexOf(x,y)];
		}

//		public void Set (Vec2i index, T val)
//		{
//			
//			data [(index.y * dim.x) + index.x] = val;
//		}
//		
//		public T Get (Vec2i index)
//		{
//			return data [(index.y * dim.x) + index.x];
//		}
	}
}

File: SW_Server/DynamicArray2D.cpp
#include "DynamicArray2D.h"


namespace SW_Server
{
	template<typename T>
	DynamicArray2D<typename T>::DynamicArray2D()
	{
	}


	template<typename T>
	DynamicArray2D<typename T>::~DynamicArray2D()
	{
	}

	/*template<typename T>
	void DynamicArray2D::AABBNonDef(out glm::ivec2 bl, out glm::ivec2 tr)
	{
		bl = glm::ivec2(dim.x, dim.y);
		tr = glm::ivec2(0, 0);

		for (int i = 0; i < dim.y; i++) {
			for (int j = 0; j < dim.x; j++) {

				if (data[indexOf(j, i)] != NULL) {

					if (j < bl.x){
						bl.x = j;
					}
					if (i < bl.y){
						bl.y = i;
					}

					if (j >= tr.x){
						tr.x = j + 1;
					}
					if (i >= tr.y){
						tr.y = i + 1;
					}
				}

			}
		}

		bl += origin;
		tr += origin;

		if (bl.x > tr.x) {
			tr.x = bl.x;
		}
		if (bl.y > tr.y) {
			tr.y = bl.y;
		}
	}

	IEnumerable<T> TryGetOutsideInside(glm::ivec2 outside, glm::ivec2 inside, int range)
	{
		int rangeT = (range * 2) + 1;

		for (int i = 0; i < rangeT; i++) {
			for (int j = 0; j < rangeT; j++) {

				glm::ivec2 temp = inside + new glm::ivec2(i - range, j - range);

				if (!((temp - outside) <= range)) {
					T temp2 = null;
					if (TryGet(temp.x, temp.y, ref temp2)) {
						yield return TryGet(temp.x, temp.y);
					}
				}
			}
		}
	}

	IEnumerable<T> WithinRange(glm::ivec2 origin, int range)
	{
		int rangeT = (range * 2) + 1;

		for (int i = 0; i < rangeT; i++) {
			for (int j = 0; j < rangeT; j++) {

				glm::ivec2 temp = origin + new glm::ivec2(i - range, j - range);

				T* temp2 = null;
				if (TryGet(temp.x, temp.y, ref temp2)) {
					yield return temp2;
				}
			}
		}
	}*/

	template<typename T>
	void DynamicArray2D<typename T>::GrowTopRight(glm::ivec2 amount)
	{
		//UnityEngine.Debug.Log("GrowTopRight: " + amount.ToString());

		glm::ivec2 newDim = dim + amount;
		T**[] newData = new T*[newDim.x * newDim.y];

		for (int i = 0; i < dim.y; i++) {
			for (int j = 0; j < dim.x; j++) {
				newData[indexOf(j, i, newDim.x)] = data[indexOf(j, i, dim.x)];
			}
		}

		data = newData;
		dim = newDim;
	}

	template<typename T>
	void DynamicArray2D<typename T>::GrowBottomLeft(glm::ivec2 amount)
	{
		//UnityEngine.Debug.Log("GrowBottomLeft: " + amount.ToString());

		glm::ivec2 newDim = dim + amount;
		T**[] newData = new T*[newDim.x * newDim.y];

		for (int i = amount.y; i < newDim.y; i++) {
			for (int j = amount.x; j < newDim.x; j++) {
				newData[indexOf(j, i, newDim.x)] = data[indexOf(j - amount.x, i - amount.y, dim.x)];
			}
		}

		origin -= amount;
		data = newData;
		dim = newDim;
	}

	template<typename T>
	bool DynamicArray2D<typename T>::Contains(int x, int y)
	{
		return (x >= origin.x && x < dim.x + origin.x) && (y >= origin.y && y < dim.y + origin.y);
	}

	template<typename T>
	bool DynamicArray2D<typename T>::Contains(glm::ivec2 index)
	{
		return (index.x >= origin.x && index.x < dim.x + origin.x) && (index.y >= origin.y && index.y < dim.y + origin.y);
	}

	template<typename T>
	int DynamicArray2D<typename T>::indexOf(glm::ivec2 index)
	{
		return (index.y * dim.x) + index.x;
	}

	template<typename T>
	int DynamicArray2D<typename T>::indexOf(glm::ivec2 index, int columns)
	{
		return (index.y * columns) + index.x;
	}

	template<typename T>
	int DynamicArray2D<typename T>::indexOf(int x, int y)
	{
		return (y * dim.x) + x;
	}

	template<typename T>
	int DynamicArray2D<typename T>::indexOf(int x, int y, int columns)
	{
		return (y * columns) + x;
	}

	template<typename T>
	void DynamicArray2D<typename T>::Set(glm::ivec2 index, T* val)
	{
		Set(index.x, index.y, val);
	}

	template<typename T>
	void DynamicArray2D<typename T>::Set(int x, int y, T* val)
	{
		if (x < origin.x) {
			if (y < origin.y) {
				//expand bottom left
				GrowBottomLeft(new glm::ivec2(origin.x - x, origin.y - y));
			}
			else if (y >= dim.y + origin.y) {
				//expand top left
				GrowTopRight(new glm::ivec2(0, y - (dim.y + origin.y) + 1));
				GrowBottomLeft(new glm::ivec2(origin.x - x, 0));
			}
			else {
				//expand left
				GrowBottomLeft(new glm::ivec2(origin.x - x, 0));
			}
		}
		else if (x >= dim.x + origin.x) {
			if (y < origin.y) {
				//expand bottom right
				GrowTopRight(new glm::ivec2(x - (dim.x + origin.x) + 1, 0));
				GrowBottomLeft(new glm::ivec2(0, origin.y - y));
			}
			else if (y >= dim.y + origin.y) {
				//expand top right
				GrowTopRight(new glm::ivec2(x - (dim.x + origin.x) + 1, y - (dim.y + origin.y) + 1));
			}
			else {
				//expand right
				GrowTopRight(new glm::ivec2(x - (dim.x + origin.x) + 1, 0));
			}
		}
		else {
			if (y < origin.y) {
				//expand bottom
				GrowBottomLeft(new glm::ivec2(0, origin.y - y));
			}
			else if (y >= dim.y + origin.y) {
				//expand top
				GrowTopRight(new glm::ivec2(0, y - (dim.y + origin.y) + 1));
			}
		}

		x -= origin.x;
		y -= origin.y;
		data[indexOf(x, y)] = val;
	}

	template<typename T>
	bool DynamicArray2D<typename T>::TryGet(int x, int y, T** val)
	{
		if (!Contains(x, y))
			return false;

		x -= origin.x;
		y -= origin.y;
		*val = data[indexOf(x, y)];

		return true;
	}

	template<typename T>
	T* DynamicArray2D<typename T>::TryGet(int x, int y)
	{
		if (!Contains(x, y))
			return null;

		x -= origin.x;
		y -= origin.y;
		return data[indexOf(x, y)];
	}

	template<typename T>
	T* DynamicArray2D<typename T>::TryGet(glm::ivec2 index)
	{
		if (!Contains(index.x, index.y))
			return null;

		index -= origin;
		return data[indexOf(index)];
	}

	template<typename T>
	T* DynamicArray2D<typename T>::Get(int x, int y)
	{
		x -= origin.x;
		y -= origin.y;
		return data[indexOf(x, y)];
	}

	template<typename T>
	T* DynamicArray2D<typename T>::Get(glm::ivec2 index)
	{
		index -= origin;
		return data[indexOf(index.x, index.y)];
	}

	template<typename T>
	T* DynamicArray2D<typename T>::Get2(int x, int y)
	{
		return data[indexOf(x, y)];
	}
}
